---
title: "avian_virus_database_pairwiseAlignment"
author: "Martin Nikolov"
date: "January 21, 2020"
output: word_document
---
# **Scanning the entire avian viral database (downloaded from NCBI Virus)**

The purpose of this script is to run a pairwise alignment between the cNRE (and cNRE reverse complement) and every single avian virus listed on NCBI Virus as of December 11, 2019. 

I had previously compiled a single FASTA file that has all these viral databases. 

The output file of this script is a set of text (.txt) files that can be converted into FASTA directly. However, these text files can be directly put into CLUSTAL's T-COFFEE multiple sequence alignment website where the multiple sequence alignment can be performed. THe CLUSTALW files produced from this alignment are then put into JalView to produce a visual multiple sequence alignment.

##Downloading necessary packages for this script
```{r eval = FALSE}

library("BiocManager")
library("BiocGenerics")
library("Biostrings")
library("GenomicRanges")
library("tidyverse")
library("splitstackshape")
library("pryr")
library("stringi")
library("dplyr")
library("ggpubr")
library("biomartr")
library("taxonomizr")

```

#Setting working directory (change this when uploading it on GitHub!)

``` {r eval = FALSE}
setwd("/Users/Martin/Desktop/UROP_WORK/My_code")
``` 


#Assigning the databases used in this script to variables

* cNRe is the cNRE sequence
* cNRE is the reverse complement of the cNRE sequence.
* HSV1_from_blast is the database of "Human herpesvirus 1 strain 17, complete genome" (GenBank accession: NC_001806.2). Downloaded November 25, 2019.
* HSV1 is the database of the "Human herpesvirus 1 strain CJ970, partial genome" (GenBank accession: JN420341.1). Downloaded (UNKNOWN DATE. THE DATE WHEN MARIANA DOWNLOADED IT).

All databases in this script are obtained from NCBI in GenBank format.

``` {r eval = FALSE}

cNRE <- readDNAStringSet("/Users/Martin/Desktop/UROP_WORK/ECRRN_full_code/ECRRN/data/raw/dna_seqs/ECRRN.fasta")
cNRE_reversecomp <- reverseComplement(cNRE)
HSV1_from_blast <-readDNAStringSet("/Users/Martin/Desktop/UROP_WORK/BLAST_analysis_data/genomes/hsv1/ncbi_genomes_2019_11_25_hsv1/GCF_000859985.2_ViralProj15217_genomic.fna")
HSV1 <- readDNAStringSet("/Users/Martin/Desktop/UROP_WORK/ECRRN_full_code/ECRRN/data/raw/virus/HSV1_genome_JN420341.fasta")

avian_viruses_full_set_ncbi <- readDNAStringSet("C:/Users/Martin/Desktop/UROP_WORK/My_code/full_list_of_avian_viral_sequences_ncbi.fasta")

```

#Pairwise alignment of each FASTA sequence in avian_viruses_full_set_ncbi against the cNRE and cNRE reverse complement (cNRE_reversecomp).

##Setting up lists and vector involved in the alignment.
A lot of these lists and vectors will be put together in the output .csv file.

* virus_name: A list of the names of the virus in avian_viruses_full_set_ncbi;
* virus_alignment_object: The objects produced from running the pairwiseAlignment function with the avian viruses and the cNRE;
* virus_alignment_score: The alignment score produced from running the pairwiseAlignment function with the avian viruses and the cNRE;
* virus_alignment_hit_sequence: The sequence within the virus that most closely aligned with the cNRE;
* virus_alignment_reversecomp_object: The objects produced from running the pairwiseAlignment function with the avian viruses and the cNRE reverse complement;
* virus_alignment_reversecomp_score: The alignment score produced from running the pairwiseAlignment function with the avian viruses and the cNRE reverse complement;
* virus_alignment_reversecomp_hit_sequence: The sequence within the virus that most closely aligned with the cNRE reverse complement;
* virus_alignment_start: The (starting) position on each viral genome where the cNRE most closely aligned. 
* virus_alignment_start_reversecomp: The (starting) position on each viral genome where the cNRE reverse complement most closely aligned. 

``` {r eval = FALSE}

virus_name <- list()
virus_alignment_object <- list()
virus_alignment_score <- list()
virus_alignment_hit_sequence <- vector()
virus_alignment_reversecomp_object <- list()
virus_alignment_reversecomp_score <- list()
virus_alignment_reversecomp_hit_sequence <- vector()

virus_alignment_start <- vector()
virus_alignment_start_reversecomp <- vector()

```


##Running pairwise alignment between each virus and the cNRE and cNRE reverse complement.

The default parameters of the pairwiseAlignment function are used and a local-global alignment is run. This means that the entire cNRE sequence is matched against part of the viral genome in each iteration.  
``` {r eval = FALSE}

for (i in 1:length(avian_viruses_full_set_ncbi)) {
  virus_name[i] <- names(avian_viruses_full_set_ncbi[i])
  virus_alignment_object[i] <- pairwiseAlignment(avian_viruses_full_set_ncbi[i], cNRE, type = "local-global")
  virus_alignment_score[i] <- score(virus_alignment_object[[i]])
  virus_alignment_hit_sequence[i] <- as.character(pattern(virus_alignment_object[[i]]))
  virus_alignment_reversecomp_object[i] <- pairwiseAlignment(avian_viruses_full_set_ncbi[i], reverseComplement(cNRE), type = "local-global")
  virus_alignment_reversecomp_score[i] <- score(virus_alignment_reversecomp_object[[i]])
  virus_alignment_reversecomp_hit_sequence[i] <- as.character(pattern(virus_alignment_reversecomp_object[[i]]))
  print(i)
}

```


##Now that the analysis is done, saving the position on each viral genome where the cNRE and cNRE reverse complement hit

``` {r eval = FALSE}

for (i in 1:length(avian_viruses_full_set_ncbi)) {
  virus_alignment_start[i] <- start(pattern(virus_alignment_object[[i]]))
  virus_alignment_start_reversecomp[i] <- start(pattern(virus_alignment_reversecomp_object[[i]]))
  print(i)
}

```


#Saving the outputs of the pairwise alignments as a data frame

``` {r eval = FALSE}

virus_alignment_dataframe_updated <- data.frame(cbind("virus name" = virus_name, 
                                                      "virus alignment object" = virus_alignment_object,
                                                      "virus alignment score" = virus_alignment_score, 
                                                      "virus alignment hit sequence" = virus_alignment_hit_sequence, 
                                                      "virus alignment start position" = virus_alignment_start, 
                                                      "virus alignment reversecomp object" = virus_alignment_reversecomp_object, 
                                                      "virus alignment reversecomp score" = virus_alignment_reversecomp_score, 
                                                      "virus alignment reversecomp hit sequence" = virus_alignment_reversecomp_hit_sequence, 
                                                      "virus alignment start position reversecomp" = virus_alignment_start_reversecomp))

```


#Convert all values in the data frame into characters

``` {r eval = FALSE}

virus_alignment_dataframe_character_updated <- mutate_all(virus_alignment_dataframe_updated, as.character)

```

#Save the data frame as a csv file
This means that the pairwise alignment results can be loaded into variables later without re-running the entire pairwise alignment.

``` {r eval = FALSE}

write.csv(cbind("virus name" = virus_alignment_dataframe_character_updated$virus.name, 
                "virus alignment score" = virus_alignment_dataframe_character_updated$virus.alignment.score, 
                "virus alignment hit sequence" = virus_alignment_dataframe_character_updated$virus.alignment.hit.sequence,
                "virus alignment start position" = virus_alignment_dataframe_character_updated$virus.alignment.start.position,
                "virus alignment reversecomp score" = virus_alignment_dataframe_character_updated$virus.alignment.reversecomp.score, 
                "virus alignment reversecomp hit sequence" = virus_alignment_dataframe_character_updated$virus.alignment.reversecomp.hit.sequence,
                "virus alignment start position reversecomp" = virus_alignment_dataframe_character_updated$virus.alignment.start.position.reversecomp),
          "virus_list_alignment_updated.csv")

```

#Extracting alignment sequences from code above to save as text file for input for JalView

##Load the data from the previously saved .csv file ("virus_list_alignment_updated_sorted.csv")

``` {r eval = FALSE}

virus_alignment_dataframe_updated_loaded <- read.csv("virus_list_alignment_updated_sorted.csv")

```

##Create a vector into which the cNRE hit sequences will be saved. 
The way the characters are written in this vector are exactly how they will be saved as a text file. For the text file to be directly convertable into a FASTA file, the way the characters are written in the vector must follow the FASTA file layout.

``` {r eval = FALSE}


output_fasta_vector <- vector()

```

##Set the first two values of the vector output_fasta_vector
This will be the name and sequence of the cNRE sequence respectively and will correspond to the first sequence in the text output file. 

``` {r eval = FALSE}
output_fasta_vector[1] <- as.character(">cNRE")
output_fasta_vector[2] <- as.character(cNRE)

```

##Filling up the rest of the values of the vector output_fasta_vector
Because this vector will later be converted into a text file and then into a FASTA sequence, all values set inside it are characters. 

``` {r eval = FALSE}

for (i in seq(from = 3, to = length(virus_alignment_dataframe_updated_loaded$virus.name), by = 2)) {
  output_fasta_vector[i] <- paste(">", as.character(virus_alignment_dataframe_updated_loaded$virus.name[(i - 1)/2]), sep = "")
  output_fasta_vector[i + 1] <- as.character(virus_alignment_dataframe_updated_loaded$virus.alignment.hit.sequence[(i - 1)/2])
  print(i)
  print(i + 1)
}

```

##Creating different output text files
These can be directly put into CLUSTAL where a multiple sequence alignment can be performed. Multiple files were created so that either many sequences can be observed or only the top few. 


``` {r eval = FALSE}

writeLines(output_fasta_vector, "output_fasta_file.txt", sep = "\n", useBytes = FALSE)
writeLines(output_fasta_vector[1:200], "output_fasta_file_1_100.txt", sep = "\n", useBytes = FALSE)
writeLines(output_fasta_vector[1:6], "output_fasta_file_1_3.txt", sep = "\n", useBytes = FALSE)
writeLines(output_fasta_vector[1:32], "output_fasta_file_1_16.txt", sep = "\n", useBytes = FALSE)

```

##**Repeating the previous steps in loading the alignment sequences and creating output text files, but this time with the cNRE reverse complement (not cNRE forward strand) virus alignments**

##Load the data from the previously saved .csv file ("virus_list_alignment_updated_reversecomp_sorted.csv")
This csv file was not created in this script. It was created by loading virus_list_alignment_updated_sorted.csv into Excel and sorting by the pairwise alignment score of the cNRE reverse complement and viruses (rather than the pairwise alignment score of the cNRE and viruses).

``` {r eval = FALSE}

virus_alignment_dataframe_updated_reversecomp_loaded <- read.csv("virus_list_alignment_updated_reversecomp_sorted.csv")

```


##Create a vector into which the cNRE *reverse complement* hit sequences will be saved. 

``` {r eval = FALSE}

output_fasta_vector_reversecomp <- vector()

```


##Set the first two values of the vector output_fasta_vector_reversecomp

``` {r eval = FALSE}

output_fasta_vector_reversecomp[1] <- as.character(">cNRE_reversecomp")
output_fasta_vector_reversecomp[2] <- as.character(reverseComplement(cNRE))

```

##Filling up the rest of the values of the vector output_fasta_vector_reversecomp

``` {r eval = FALSE}

for (i in seq(from = 3, to = length(virus_alignment_dataframe_updated_reversecomp_loaded$virus.name), by = 2)) {
  output_fasta_vector_reversecomp[i] <- paste(">", as.character(virus_alignment_dataframe_updated_reversecomp_loaded$virus.name[(i - 1)/2]), sep = "")
  output_fasta_vector_reversecomp[i + 1] <- as.character(virus_alignment_dataframe_updated_reversecomp_loaded$virus.alignment.reversecomp.hit.sequence[(i - 1)/2])
  print(i)
  print(i + 1)
}

```


##Creating different output text files

``` {r eval = FALSE}

writeLines(output_fasta_vector_reversecomp, "output_fasta_filer_reversecomp.txt", sep = "\n", useBytes = FALSE)
writeLines(output_fasta_vector_reversecomp[1:200], "output_fasta_filer_reversecomp_1_100.txt", sep = "\n", useBytes = FALSE)
writeLines(output_fasta_vector_reversecomp[1:6], "output_fasta_filer_reversecomp_1_3.txt", sep = "\n", useBytes = FALSE)
writeLines(output_fasta_vector_reversecomp[1:32], "output_fasta_filer_reversecomp_1_16.txt", sep = "\n", useBytes = FALSE)
writeLines(output_fasta_vector_reversecomp[1:38], "output_fasta_filer_reversecomp_1_19.txt", sep = "\n", useBytes = FALSE)
writeLines(output_fasta_vector_reversecomp[1:10], "output_fasta_filer_reversecomp_1_5.txt", sep = "\n", useBytes = FALSE)

```



##Running pairwise alignment between the two top hits against the cNRE reverse complement (Psittacus erithacus timneh papillomavirus and Falconid herpesvirus 1 strain S-18) to manually produce Jalview visual output (first reverse complementing the hit sequences so that we show an alignment with the cNRE, not cNRE reverse complement)

``` {r eval = FALSE}

First_hit_reversecomp_reversecomp <- reverseComplement(DNAStringSet(virus_alignment_dataframe_updated_reversecomp_loaded$virus.alignment.reversecomp.hit.sequence[[1]]))
Second_hit_reversecomp_reversecomp <- reverseComplement(DNAStringSet(virus_alignment_dataframe_updated_reversecomp_loaded$virus.alignment.reversecomp.hit.sequence[[3]]))

```



``` {r eval = FALSE}

first_alignment <- pairwiseAlignment(cNRE, First_hit_reversecomp_reversecomp, type = "global")
second_alignment <- pairwiseAlignment(cNRE, Second_hit_reversecomp_reversecomp, type = "global")

```


``` {r eval = FALSE}

writeLines(c(">cNRE", 
             as.character(pattern(first_alignment)), 
             paste(">", as.character(virus_alignment_dataframe_updated_reversecomp_loaded$virus.name[[1]]), sep = ""), 
             as.character(subject(first_alignment))),
           "first_reversecomp_alignment.txt")

writeLines(c(">cNRE", 
             as.character(pattern(second_alignment)), 
             paste(">", as.character(virus_alignment_dataframe_updated_reversecomp_loaded$virus.name[[3]]), sep = ""), 
             as.character(subject(second_alignment))),
           "second_reversecomp_alignment.txt")

```


